File: clients/algoliasearch-client-csharp/algoliasearch/Transport/StatefulHost.cs
using System;

namespace Algolia.Search.Transport;

/// <summary>
/// Algolia's stateful host
/// </summary>
public class StatefulHost
{
  /// <summary>
  /// Url endpoint without the scheme, without the port
  /// </summary>
  public string Url { get; set; }

  /// <summary>
  /// Scheme of the URL
  /// </summary>
  public HttpScheme Scheme { get; set; } = HttpScheme.Https;

  /// <summary>
  /// Port of the URL (Optional)
  /// </summary>
  public int Port { get; set; } = -1;

  /// <summary>
  /// Is the host up or not
  /// </summary>
  public bool Up { get; set; } = true;

  /// <summary>
  /// Retry count
  /// </summary>
  public int RetryCount { get; set; }

  /// <summary>
  /// Last time the host has been used
  /// </summary>
  public DateTime LastUse { get; set; } = DateTime.UtcNow;

  /// <summary>
  /// Calltype accepted by the host
  /// </summary>
  public CallType Accept { get; set; }
}

public enum HttpScheme
{
  /// <summary>
  /// Http
  /// </summary>
  Http,
  /// <summary>
  /// Https
  /// </summary>
  Https
}

/// <summary>
/// https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/enumeration-types
/// Binary enums beware when adding new values
/// </summary>
[Flags]
public enum CallType
{
  /// <summary>
  /// Read Call
  /// </summary>
  Read = 1,

  /// <summary>
  /// Write Call
  /// </summary>
  Write = 2
}

File: clients/algoliasearch-client-ruby/lib/algolia/transport/stateful_host.rb
module Algolia
  module Transport
    class StatefulHost
      include CallType

      attr_reader :url, :protocol, :accept, :port
      attr_accessor :last_use, :retry_count, :up

      # @param url [String] host url
      # @option options [binary] :accept accept type flag
      # @option options [DateTime] :last_use last usage date
      # @option options [Integer] :retry_count number of retries
      # @option options [Boolean] :up host status
      #
      def initialize(url, opts = {})
        @url         = url
        @protocol    = opts[:protocol] || 'https://'
        @port        = opts[:port]
        @accept      = opts[:accept] || (READ | WRITE)
        @last_use    = opts[:last_use] || Time.now.utc
        @retry_count = opts[:retry_count] || 0
        @up          = opts.key?(:up) ? opts[:up] : true
      end
    end
  end
end
